### PR TITLE
Change facility

### PR DESCRIPTION
**Pull request by saket**

### CODE REVIEW

File: app/src/main/java/org/simple/clinic/facility/FacilityRepository.kt [hunk 1]
```diff
@@ -29,14 +28,18 @@ class FacilityRepository @Inject constructor(
...
+  fun associateUserWithFacility(userSession: UserSession, facilityId: UUID): Completable {
+    return requireUser(userSession)
+        .take(1)
+        .flatMapCompletable {
+          Completable.fromAction {
+            userFacilityMappingDao.insertOrUpdate(it, listOf(facilityId), newCurrentFacilityUuid = facilityId)
           }
         }
-        .map { (user) -> user }
+  }
```
Comment (saket) [inline app/src/main/java/org/simple/clinic/facility/FacilityRepository.kt]: @vinaysshenoy what do you think about `associateUserWithFacility()`?

File: app/src/main/res/drawable/ic_keyboard_arrow_down_20dp.xml [hunk 2]
```diff
@@ -0,0 +1,11 @@
+<vector xmlns:android="http://schemas.android.com/apk/res/android"
```
Comment (vinaysshenoy) [inline app/src/main/res/drawable/ic_keyboard_arrow_down_20dp.xml]: Github needs a way to preview vector drawables. :D

File: app/src/main/java/org/simple/clinic/facility/FacilityRepository.kt (hunk 1 above)
Comment (saket) [reply app/src/main/java/org/simple/clinic/facility/FacilityRepository.kt]: `setActiveFacilityForUser()` sounds like it has to be used when an existing facility mapping already exists.

File: app/src/main/java/org/simple/clinic/facility/FacilityRepository.kt [hunk 3]
```diff
@@ -29,14 +28,18 @@ class FacilityRepository @Inject constructor(
     }
   }
 
-  fun currentFacility(userSession: UserSession): Observable<Facility> {
-    return userSession.loggedInUser()
-        .doOnNext {
-          if (it === None) {
-            throw AssertionError("User isn't logged in yet")
+  fun associateUserWithFacility(userSession: UserSession, facilityId: UUID): Completable {
```
[the remark this answers is not in the document]
Comment (saket) [reply app/src/main/java/org/simple/clinic/facility/FacilityRepository.kt]: It's the same question as DAOs accessing each other

[the remark this answers is not in the document]
Comment (saket) [reply app/src/main/java/org/simple/clinic/facility/FacilityRepository.kt]: `currentFacility()` is called from many places, so I felt a convenience function would be good.

File: app/src/main/java/org/simple/clinic/facility/change/FacilityChangeScreen.kt [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        .compose(controller)
+        .observeOn(AndroidSchedulers.mainThread())
+        .takeUntil(RxView.detaches(this))
+        .subscribe { uiChange -> uiChange(this) }
+
+    toolbar.setNavigationOnClickListener {
+      screenRouter.pop()
+    }
+
+    facilityRecyclerView.layoutManager = LinearLayoutManager(context)
```
[the remark this answers is not in the document]
Comment (saket) [reply app/src/main/java/org/simple/clinic/facility/change/FacilityChangeScreen.kt]: Uhm… I don't think I care :p

File: app/src/main/java/org/simple/clinic/facility/FacilityRepository.kt (hunk 1 above)
Comment (saket) [reply app/src/main/java/org/simple/clinic/facility/FacilityRepository.kt]: 🙂 👍

File: app/src/main/res/drawable/ic_keyboard_arrow_down_20dp.xml (hunk 2 above)
Comment (saket) [reply app/src/main/res/drawable/ic_keyboard_arrow_down_20dp.xml]: Can we somehow tag Chris Wanstrath here

File: app/src/main/res/layout/screen_app_lock.xml [hunk 5]
```diff
@@ -5,17 +5,67 @@
...
+        android:layout_width="0dp"
+        android:layout_height="wrap_content"
+        android:layout_weight="1"
+        android:ellipsize="end"
+        android:maxLines="1"
+        android:textColor="@color/white"
+        android:textSize="@dimen/textsize_20"
+        tools:text="A really long person name" />
+
+      <android.support.v7.widget.AppCompatButton
```
[the remark this answers is not in the document]
Comment (saket) [reply app/src/main/res/layout/screen_app_lock.xml]: I was earlier using `AppCompatButton` to override `stateListAnimator`, but that's no longer needed. I will fix this, thanks for noticing.

File: app/src/main/java/org/simple/clinic/facility/change/FacilityChangeScreenController.kt [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+  override fun apply(events: Observable<UiEvent>): ObservableSource<UiChange> {
+    val replayedEvents = events.compose(ReportAnalyticsEvents()).replay().refCount()
+
+    return Observable.mergeArray(
+        showFacilities(replayedEvents),
+        changeFacilityAndExit(replayedEvents))
+  }
+
+  private fun showFacilities(events: Observable<UiEvent>): Observable<UiChange> {
+    val facilitiesStream = events
```
[the remark this answers is not in the document]
Comment (saket) [reply app/src/main/java/org/simple/clinic/facility/change/FacilityChangeScreenController.kt]: Whoops, inlining this

File: app/src/main/java/org/simple/clinic/facility/FacilityRepository.kt [hunk 7]
```diff
@@ -29,14 +28,18 @@ class FacilityRepository @Inject constructor(
...
+        .take(1)
+        .flatMapCompletable {
+          Completable.fromAction {
+            userFacilityMappingDao.insertOrUpdate(it, listOf(facilityId), newCurrentFacilityUuid = facilityId)
           }
         }
-        .map { (user) -> user }
+  }
+
+  fun currentFacility(userSession: UserSession): Observable<Facility> {
```
[the remark this answers is not in the document]
Comment (saket) [reply app/src/main/java/org/simple/clinic/facility/FacilityRepository.kt]: Done. I'll change the usages in other places in a future commit.

**Pull request merged**